test/client/dio_interceptor_test.dart

```diff
@@ -0,0 +1,354 @@
+import 'dart:convert';
+import 'dart:io';
+import 'dart:math';
+import 'dart:typed_data';
+
+import 'package:cryptography/cryptography.dart';
+import 'package:cryptography/dart.dart';
+import 'package:dio/dio.dart';
+import 'package:http_mock_adapter/http_mock_adapter.dart';
+import 'package:requests_signature_dart/src/client/requests_signature_options.dart';
+import 'package:test/test.dart';
+import 'package:requests_signature_dart/requests_signature_dart.dart';
+import 'package:uuid/uuid.dart';
+
+void main() {
+  group('Requests Signature Interceptor Test', () {
+    final testClientId = 'test_client_id';
+    final testClientSecret = 'test_client_secret';
+    final headerName = 'X-Signature';
+    final signaturePattern = '{ClientId}:{Nonce}:{Timestamp}:{SignatureBody}';
+    final baseEndpoint = 'https://google.ca/';
+    final emptyDataInjection = '{}'; //Injecting empty data as String
+    final Uuid _uuid = const Uuid();
+
+    late Dio dio;
+    late int clockskew; // in milliseconds
+
+    setUp(() {
+      dio = Dio(BaseOptions(
+        validateStatus: (status) =>
+            true, // We need to set this to avoid unnecessary throwing exception when we expect an error status.
+      ));
+      // Setting random clockskew so tests aren't always relying on the same value.
+      clockskew = Random().nextInt(2000) + 1000;
+    });
+
+    tearDown(() {
+      dio.close();
+    });
+
+    test('Interceptor adds signature header to GET request.', () async {
+      // Arrange
+      final options = RequestsSignatureOptions(
+        clientId: testClientId,
+        clientSecret: testClientSecret,
+        headerName: headerName,
+        signaturePattern: signaturePattern,
+      );
+
+      final dioAdapter = DioAdapter(dio: dio);
+      dio.httpClientAdapter = dioAdapter;
+
+      final interceptor = RequestsSignatureInterceptor(options, dio);
+
+      dio.interceptors.add(interceptor);
+
+      var mockServer = MockServerValidator(
+        headerName,
+        clockskew,
+        signaturePattern,
+      );
+
+      // Register the mock response
+      dioAdapter.onGet(
+        baseEndpoint,
+        (server) {
+          var headers = {
+            "date": [DateTime.now().toString()]
+          };
+
+          server.replyCallbackAsync(
+            HttpStatus.ok,
+            (request) async {
+              var isValid =
+                  await mockServer.validateSignature(request, testClientSecret);
+
+              if (isValid) {
+                return emptyDataInjection;
+              } else {
+                // The status code can only be set while defining the mock. Therefore, we're throwing an exception to simulate an unexpected status code.
+                throw "The signature is invalid!";
```

Comment:
But the server needs to return 4001 when the signature is invalid.

Reply:
Right. However, the status code (as a parameter in the method `replyCallbackAsync`) can only be mocked before the signature validation logic is executed. That's why I went with the exception instead.

Reply:
Or I should say, it seems impossible with what we have to return a different status code. You can only set it once as a parameter of `replyCallbackAsync`.